### PR TITLE
chore: Move the widget config to widget class

### PR DESCRIPTION
**Pull request by sbalaji1192**
## Description
- Remove the config objects from widget and config maps from the widget factory.
- Introduce methods in widget development API to dynamically fetch this items.
- freeze the widget configuration.

#### PR fixes following issue(s)
Fixes https://github.com/appsmithorg/appsmith/issues/26008
> if no issue exists, please create an issue and ask the maintainers about this first
>
>
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Chore (housekeeping or task changes that don't impact user perception)
- This change requires a documentation update
>
>
>
- [x] Manual
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [x] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [x] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/src/WidgetProvider/factory/index.tsx [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+import type { RegisteredWidgetFeatures } from "../../utils/WidgetFeatures";
+import memo from "micro-memoize";
+// import { WIDGETS_COUNT } from "widgets";
+import type { AnyFn } from "micro-memoize/src/types";
+
+type WidgetDerivedPropertyType = any;
+export type DerivedPropertiesMap = Record<string, string>;
+export type WidgetType = (typeof WidgetFactory.widgetTypes)[number];
+
+function memoize(fn: AnyFn) {
```
Comment (riodeuno) [inline app/client/src/WidgetProvider/factory/index.tsx:37]: @sbalaji1192 Should we make this into a util function? Seems like it is generic enough to be used across the codebase. Also, can you describe why we used this function over other libraries that already exist in the system?

File: app/client/src/WidgetProvider/factory/index.tsx [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+      isDeprecated: !!config.isDeprecated,
+      replacement: config.replacement,
+      displayName: config.name,
+      key: generateReactKey(),
+      iconSVG: config.iconSVG,
+      isCanvas: config.isCanvas,
+      needsHeightForContent: config.needsHeightForContent,
+    };
+
+    WidgetFactory.widgetConfigMap.set(widget.type, Object.freeze(_config));
```
Comment (riodeuno) [inline app/client/src/WidgetProvider/factory/index.tsx:106]: Are we getting rid of the WidgetConfigs reducer?

File: app/client/src/WidgetProvider/factory/index.tsx [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+export type DerivedPropertiesMap = Record<string, string>;
+export type WidgetType = (typeof WidgetFactory.widgetTypes)[number];
+
+function memoize(fn: AnyFn) {
+  return memo(fn, {
+    maxSize: 100,
+  });
+}
+
+class WidgetFactory {
```
Comment (riodeuno) [inline app/client/src/WidgetProvider/factory/index.tsx:43]: @sbalaji1192 As this acts as a in memory data store. Should we call Object.freeze on these configs? I believe that immutable data makes the system more stable, especially in a language like JS.

File: app/client/src/widgets/InputWidget/widget/index.tsx [hunk 4]
```diff
@@ -126,6 +131,74 @@ class InputWidget extends BaseWidget<InputWidgetProps, WidgetState> {
...
+            propertyPath: "defaultText",
+            propertyValue: propValueMap.data,
+            isDynamicPropertyPath: true,
+          },
+        ];
+      },
+    };
+  }
+
+  static getAutoLayoutConfig() {
```
Comment (berzerkeer) [inline app/client/src/widgets/InputWidget/widget/index.tsx:186]: Previous change doesn't have auto layout config. So i think we should remove this.

File: app/client/src/widgets/CheckboxGroupWidget/widget/index.tsx [hunk 5]
```diff
@@ -59,6 +62,80 @@ class CheckboxGroupWidget extends BaseWidget<
...
+  static getFeatures() {
+    return {
+      dynamicHeight: {
+        sectionIndex: 3,
+        active: true,
+      },
+    };
+  }
+
+  static getAutolayoutConfif() {
```
Comment (keyurparalkar) [inline app/client/src/widgets/CheckboxGroupWidget/widget/index.tsx:110]: @sbalaji1192 this should be` getAutoLayoutConfig`

File: app/client/src/widgets/MenuButtonWidget/widget/index.tsx [hunk 6]
```diff
@@ -5,16 +5,100 @@ import { isArray, orderBy } from "lodash";
...
+        },
+      },
+      rows: 4,
+      columns: 16,
+      widgetName: "MenuButton",
+      version: 1,
+    };
+  }
+
+  static getAutoLayoutconfig() {
```
Comment (berzerkeer) [inline app/client/src/widgets/MenuButtonWidget/widget/index.tsx:74]: `getAutoLayoutConfig`  , `C` should be caps.

**Comment by laveena-en** [comment]
@sbalaji1192 onClick event isn’t getting executed with the Modal widget when it is renamed. Can you please check: https://www.loom.com/share/03199895f5d44de582041c300e4502c3?sid=cc056b66-396f-49aa-a24c-02c7c798b6ac

**Comment by sbalaji1192** [comment]
@laveena-en This is an existing issue, we can repro this on the release env as well. Can you please check once and create a new issue?

**Review by keyurparalkar — APPROVED** [review]

**Review by berzerkeer — APPROVED** [review]

**Review by keyurparalkar — APPROVED** [review]

**Pull request merged**